Raw interconnection queue data

jdangerx opened the pull request:
Closes #305.

This adds a `get_raw_interconnection_queue` method to all `ISOBase` subclasses that outputs a `BinaryIO`. This allows users like Catalyst to archive the raw data, including data that we don't know how to interpret within GridStatus yet.

I parametrized the `test_interconnection_queue` test so that a failure in one ISO doesn't fail the whole thing. This helps us isolate test failures. I also added an assertion that the queue data is non-empty to slightly strengthen the tests, and fixed the NYISO column mapping code to handle a change in their data format.

kmax12 commented:
thanks for getting the ball rolling on this. 

i only took a very quick look, but i suspect the issue has to do with NYISO changing their data format. you may just have to update the nyiso specific code until it works again

kmax12 commented:
better testing for interconnection queue would be appreciated and i suspect would make them more likely to be used by others. i implemented these initially, but havent been using these scrapers on gridstatus.io, so they haven't been maintained like everything else.

jdangerx commented:
There were a few column name/order changes so I added an explicit remap and NYISO appears to be working again!

In terms of better testing/assertions, I made some incremental improvements (parametrization, asserting non-empty) but don't have the domain knowledge to really make other meaningful improvements right now - that might make sense as a separate PR.

jdangerx commented on gridstatus/nyiso.py:
```diff
@@ -377,7 +386,35 @@ def get_interconnection_queue(self, verbose=False):
+            ("SGIA Tender Date", ""): "SGIA Tender Date",
+            ("CY Complete Date", ""): "CY Complete Date",
+            ("Proposed Initial-Sync Date", ""): "Proposed Initial-Sync Date",
+            ("Proposed", " In-Service"): "Proposed In-Service Date",
```
What's the relationship between the `"Proposed In-Service Date"` column and the `"Proposed  In-Service"` column that we create below with two spaces?

kmax12 replied:
i think the double space is a typo! you can fix that

if i remember my own intentions, the goal was to just remove the "I/S" or "IS" values so the type of the column would be a datetime. have the string would cause a mixed typed

jdangerx commented on pyproject.toml:
```diff
@@ -46,7 +46,7 @@ test = [
 ]
 dev = [
-    "ruff == 0.0.202",
+    "ruff >= 0.0.291",
```
VS Code was complaining about `ruff == 0.0.202` being too old for the Ruff extension, so I updated it here - if you have any specific reason to keep it pinned to such an old version, we can definitely do that. Current version is 1.9, FWIW.

jdangerx commented on gridstatus/nyiso.py:
```diff
@@ -63,7 +66,9 @@ def _parse_status(row):
                 row["Status"] = row["Status"][
                     row["Status"].index(STATE_CHANGE)
-                    + len(STATE_CHANGE) : -len(" state.**")
+                    + len(STATE_CHANGE) : -len(
```
Maybe an artifact of updating `ruff`?

kmax12 replied:
must be

jdangerx commented on gridstatus/nyiso.py:
```diff
@@ -377,7 +386,35 @@ def get_interconnection_queue(self, verbose=False):
         ):
             active = active.drop(columns=["SGIA Tender Date"])
-        completed.columns = active.columns
+        completed_colnames_map = {
```
The original code here assumed that the columns were conceptually the same between `active` and `columns`.

After taking a look I saw two differences:

1. There were three `"Proposed COD.X"` columns in `completed` that were not in `active`.
2. The order of `"Proposed Initial-Sync Date"` and `"Proposed In-Service Date"` were flipped between the two.

So I made the mapping explicit.

WillKoehrsen replied:
Yeah, it sounds like others have noticed issues between the active sheet and other sheets so thanks for taking care of this (#323).

jdangerx commented:
@kmax12 I made a bunch of comments but I think this is pretty much ready for review!

kmax12 commented:
Thanks @jdangerx. I’ll review as I have time. Might take a week or so but will try to be faster

kmax12 commented on gridstatus/tests/test_interconnection_queue.py:
```diff
@@ -5,6 +5,9 @@
 @pytest.mark.slow
-def test_get_interconnection_queue_all():
+@pytest.mark.parametrize("iso_class", gridstatus.utils.all_isos)
+def test_get_interconnection_queue_all(iso_class):
```
we recently added `IESO` to [`all_isos`](https://github.com/kmax12/gridstatus/blob/c12055616f3da4e34ddde181469cd8d976133a8e/gridstatus/__init__.py#L31). there is an interconnection queue [here](https://www.ieso.ca/en/Sector-Participants/Connection-Process/Application-Status), but we haven't implemented it. let's just make sure this test can handle ieso not being implemented

kmax12 reviewed (commented):
sorry for the long delay on this. this is a nice refactor to make the raw queue data available for archiving. 

just a tiny bit left on nyiso and we can merge this. 

thanks again!

jdangerx commented:
I fixed the double space typo and updated the tests to not fail on IESO. I'm not quite sure if you want me to merge the `Proposed In-Service` and `Proposed In-Service Date` columns or not, but I can do that and strip out the `I/S` and `IS` values if that is the right thing to do!

kmax12 commented:
> I fixed the double space typo and updated the tests to not fail on IESO. I'm not quite sure if you want me to merge the `Proposed In-Service` and `Proposed In-Service Date` columns or not, but I can do that and strip out the `I/S` and `IS` values if that is the right thing to do!

yes, i think they can be merged. i think the `I/S` value that shows up for completed projects is just saying that they are in fact "in service". this column can be null for completed project

kmax12 commented on gridstatus/nyiso.py:
```diff
@@ -441,8 +478,8 @@ def get_interconnection_queue(self, verbose=False):
         )
-        queue["Proposed  In-Service"] = pd.to_datetime(
-            queue["Proposed  In-Service"],
+        queue["Proposed In-Service"] = pd.to_datetime(
```
should this be `"Proposed In-Service Date"` since that is the value in `active`?

kmax12 commented on gridstatus/nyiso.py:
```diff
@@ -472,7 +509,7 @@ def get_interconnection_queue(self, verbose=False):
 
         extra_columns = [
-            "Proposed  In-Service",
+            "Proposed In-Service",
```
same comment as above

should this be "Proposed In-Service Date" since that is the value in active?

WillKoehrsen reviewed (approved):
@jdangerx The tests pass and all the changes look good. Is this ready to merge?

jdangerx commented:
Yeah, it's ready for merge! There was a merge conflict which I just dealt with - hopefully CI continues to pass 😅

WillKoehrsen commented:
@jdangerx It looks like there's a linting issue. Running `make lint-fix` then committing changes should solve that.

WillKoehrsen commented:
It looks like there is [one test failure](https://github.com/kmax12/gridstatus/actions/runs/7788746640/job/21239283143?pr=307#step:6:2458) on `TestNYISO.test_get_interconnection_queue_handles_new_file`. That was a test I added in #329 to handle new NYISO interconnection data. I think we just need to make sure the `"Interconnection Location"` column is handled correctly for NYISO.

jdangerx commented:
Ah, yeah, looks like you renamed `Points of Interconnection` in the active sheet, which propagated downstream in the old code but needs to be manually updated for the completed sheet now. I think everything should pass now 🤞 Thanks for your patience!

WillKoehrsen commented:
All the tests pass now, so I'll go ahead and merge. Thanks @jdangerx!
